agent/connect/ca/provider_aws.go

```diff
@@ -0,0 +1,700 @@
+package ca
+
+import (
+	"bytes"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"log"
+	"sync/atomic"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/acmpca"
+	"github.com/mitchellh/mapstructure"
+
+	"github.com/hashicorp/consul/agent/connect"
+	"github.com/hashicorp/consul/agent/structs"
+)
+
+const (
+	// RootTemplateARN is the AWS-defined template we need to use when issuing a
+	// root cert.
+	RootTemplateARN = "arn:aws:acm-pca:::template/RootCACertificate/V1"
+
+	// IntermediateTemplateARN is the AWS-defined template we need to use when
+	// issuing an intermediate cert.
+	IntermediateTemplateARN = "arn:aws:acm-pca:::template/SubordinateCACertificate_PathLen0/V1"
+
+	// LeafTemplateARN is the AWS-defined template we need to use when issuing a
+	// leaf cert.
+	LeafTemplateARN = "arn:aws:acm-pca:::template/EndEntityCertificate/V1"
+
+	// RootTTL is the validity duration for root certs we create.
+	RootTTL = 5 * 365 * 24 * time.Hour
+
+	// IntermediateTTL is the validity duration for the intermediate certs we
+	// create.
+	IntermediateTTL = 1 * 365 * 24 * time.Hour
+)
+
+// AWSProvider implements Provider for AWS ACM PCA
+type AWSProvider struct {
+	stopped uint32 // atomically accessed, at start to prevent alignment issues
+	stopCh  chan struct{}
+
+	config          *structs.AWSCAProviderConfig
+	session         *session.Session
+	client          *acmpca.ACMPCA
+	isPrimary       bool
+	datacenter      string
+	clusterID       string
+	arn             string
+	arnChecked      bool
+	caCreated       bool
+	rootPEM         string
+	intermediatePEM string
+	logger          *log.Logger
+}
+
+// SetLogger implements NeedsLogger
+func (a *AWSProvider) SetLogger(l *log.Logger) {
+	a.logger = l
+}
+
+// Configure implements Provider
+func (a *AWSProvider) Configure(cfg ProviderConfig) error {
+	config, err := ParseAWSCAConfig(cfg.RawConfig)
+	if err != nil {
+		return err
+	}
+
+	// We only support setting IAM credentials through the normal methods ENV,
```

Comment:
Would it be possible to have the state store diverge if all of the servers had different IAM credentials that implied that they were manipulating CAs in wildly different regions from each other?

Reply:
No it would just break - the actual state stored is the ARN of a specific CA which only exists in one region so if servers have different IAM configs then some just would fail to load it.